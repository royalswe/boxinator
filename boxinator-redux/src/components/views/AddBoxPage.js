/**
 * TODO:: Form and POST validation after backend validation is done
 * exports AddBoxPage class only for testing purpose
 */

import React from 'react';
import classnames from 'classnames';
import { connect } from 'react-redux';
import { createItem } from '../../actions/actions';

export class AddBoxPage extends React.Component {

    state = {
        name: '',
        color: '',
        weight: '',
        country: '',
        errors: {}
    };

    handleChange = (e) => {
        /**
         * If there is any validation errors,
         * validation errors disappears when isvalid
         */
        if (!!this.state.errors[e.target.name]) {
          let errors = Object.assign({}, this.state.errors);
          delete errors[e.target.name];
          this.setState({
            [e.target.name]: e.target.value,
            errors
          });
        } else {
          this.setState({ [e.target.name]: e.target.value });
        }
      }

    handleSubmit = (e) => {
        e.preventDefault();

        /**
         * Validate form and prevent submit if is not valid
         */
        let errors = {};
        if(this.state.name === '') errors.name = "Name can't be empty";

        this.setState({errors});
        const isValid = Object.keys(errors).length === 0;
        if(isValid){ // if form dont have errors
            const { name, color, weight, country } = this.state;
            this.props.createItem({name, color, weight, country})
        }
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <h2>Add new box</h2>

                <div className={classnames('field', { error: !!this.state.errors.name})}>
                    <label htmlFor="name">Name</label>
                    <input
                        name="name"
                        value={this.state.name}
                        onChange={this.handleChange}
                        id="name"
                    />
                </div>
                <span>{this.state.errors.name}</span>
                <div>
                    <label htmlFor="weight">Weight</label>
                    <input
                        name="weight"
                        value={this.state.weight}
                        onChange={this.handleChange}
                        id="weight"
                    />
                </div>
                <div>
                    <label htmlFor="color">Color</label>
                    <input
                        name="color"
                        value={this.state.color}
                        onChange={this.handleChange}
                        id="color"
                    />
                </div>
                <div>
                    <label htmlFor="country">Country</label>
                    <input
                        name="country"
                        value={this.state.country}
                        onChange={this.handleChange}
                        id="country"
                    />
                </div>
                <div>
                    <button>Save</button>
                </div>
            </form>
        )
    }
}

export default connect(null, { createItem })(AddBoxPage);