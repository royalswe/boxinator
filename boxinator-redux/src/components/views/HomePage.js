import React from 'react';

class HomePage extends React.Component {

    render() {
        return (
            <div>
                <h2>Welcome to Boxinator!</h2>
                <p>This is a simple application build with React and Redux</p>
            </div>
        )
    }
}

export default HomePage;