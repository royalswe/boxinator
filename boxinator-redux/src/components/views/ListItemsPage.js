import React from 'react';
import { connect } from 'react-redux';
import ItemList from '../ItemList';
import PropTypes from 'prop-types';
import { fetchItems } from '../../actions/actions';


class ListItemsPage extends React.Component {
  componentDidMount(){
    this.props.fetchItems();
  }
  render() {
    console.log(this.props.items);
    
    return (
      <div>
        <h1>Box List</h1>
        <ItemList items={this.props.items} />
      </div>
    );
  }
  
}

ListItemsPage.propTypes = {
  items: PropTypes.array.isRequired,
  fetchItems: PropTypes.array.isRequired
}

function mapStateToProps(state) {
  return {
    items: state.items
  }
}

export default connect(mapStateToProps, { fetchItems })(ListItemsPage);

