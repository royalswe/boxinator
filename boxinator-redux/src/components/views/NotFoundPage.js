import React from 'react';

class HomePage extends React.Component {

    render() {
        // Just a simple 404 page
        return (
            <div>
                <h2>404</h2> 
                <h4>Could not find the page <code>{window.location.pathname}</code></h4>
            </div>
        )
    }
}

export default HomePage;