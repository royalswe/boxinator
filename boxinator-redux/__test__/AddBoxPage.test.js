import React from 'react';
import AddBoxPageRedux from '../src/components/views/AddBoxPage';
import { AddBoxPage } from '../src/components/views/AddBoxPage';
import { shallow, mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('<ListItemsPage /> should render box items', () => {
    /**
    * middleware function. This is a mockup for the redux-thunk.
    */
    const thunk = ({ dispatch, getState }) => next => action => {
    if (typeof action === 'function') {
        return action(dispatch, getState)
    }
        return next(action)
    }
    /**
     * fake the redux methods getState, dispatch, and next
     */
    const create = () => {
    const store = {
        getState: jest.fn(() => ({})),
        dispatch: jest.fn()
    };
        const next = jest.fn()
        const invoke = (action) => thunk(store)(next)(action)
        return {store, next, invoke}
    };

    
    it('render the form in <AddBoxPage />', () => {
        const wrapper = shallow(<AddBoxPage />);
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

    it('passes dispatch and getState to addBoxPage component', () => {
        const { store, invoke } = create()
        invoke((dispatch, getState) => {
          dispatch('ADD_ITEM_SUCCESS')
          getState();
        })

        const wrapper = shallow(<AddBoxPageRedux store={store} />);
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });
});