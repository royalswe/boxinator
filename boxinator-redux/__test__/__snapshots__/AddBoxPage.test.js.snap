// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItemsPage /> should render box items passes dispatch and getState to addBoxPage component 1`] = `
<AddBoxPage
  createItem={[Function]}
  store={
    Object {
      "dispatch": [MockFunction] {
        "calls": Array [
          Array [
            "ADD_ITEM_SUCCESS",
          ],
        ],
      },
      "getState": [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
      },
    }
  }
/>
`;

exports[`<ListItemsPage /> should render box items render the form in <AddBoxPage /> 1`] = `
<div>
  <form
    className="add-box"
    onSubmit={[Function]}
  >
    <h2>
      Add new box
    </h2>
    <span
      className="error-text"
    />
    <div>
      <label
        htmlFor="name"
      >
        Name
      </label>
      <input
        id="name"
        name="name"
        onChange={[Function]}
        required={true}
        value=""
      />
    </div>
    <span
      className="error-text"
    />
    <div>
      <label
        htmlFor="weight"
      >
        Weight
      </label>
      <input
        id="weight"
        name="weight"
        onChange={[Function]}
        required={true}
        type="number"
        value=""
      />
    </div>
    <span
      className="error-text"
    />
    <div>
      <label
        htmlFor="country"
      >
        Destination Country
      </label>
      <select
        name="country"
        onChange={[Function]}
        value="Sweden"
      >
        <option
          key="Sweden"
          value="Sweden"
        >
          Sweden
        </option>
        <option
          key="China"
          value="China"
        >
          China
        </option>
        <option
          key="Brazil"
          value="Brazil"
        >
          Brazil
        </option>
        <option
          key="Australia"
          value="Australia"
        >
          Australia
        </option>
      </select>
    </div>
    <span
      className="error-text"
    />
    <div>
      <div>
        Box Color
      </div>
      <div
        onClick={[Function]}
        style={
          Object {
            "MozBorderRadius": "1px",
            "MozBoxShadow": "0 0 0 1px rgba(0,0,0,.1)",
            "OBorderRadius": "1px",
            "OBoxShadow": "0 0 0 1px rgba(0,0,0,.1)",
            "WebkitBorderRadius": "1px",
            "WebkitBoxShadow": "0 0 0 1px rgba(0,0,0,.1)",
            "background": "#fff",
            "borderRadius": "1px",
            "boxShadow": "0 0 0 1px rgba(0,0,0,.1)",
            "cursor": "pointer",
            "display": "inline-block",
            "msBorderRadius": "1px",
            "msBoxShadow": "0 0 0 1px rgba(0,0,0,.1)",
            "padding": "5px",
          }
        }
      >
        <div
          style={
            Object {
              "MozBorderRadius": "2px",
              "OBorderRadius": "2px",
              "WebkitBorderRadius": "2px",
              "background": "rgb(241, 112, 0)",
              "borderRadius": "2px",
              "height": "14px",
              "msBorderRadius": "2px",
              "width": "36px",
            }
          }
        />
      </div>
    </div>
    <div>
      <button>
        Save
      </button>
    </div>
  </form>
</div>
`;
